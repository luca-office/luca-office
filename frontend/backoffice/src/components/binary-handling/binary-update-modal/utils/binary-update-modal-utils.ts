import {UploadFileType as FileType} from "shared/enums"
import {LucaI18nLangKey} from "shared/translations"

export const getDeleteButtonKey = (deleteButtonKey: LucaI18nLangKey | undefined, fileType: FileType) => {
  if (deleteButtonKey) {
    return deleteButtonKey
  } else {
    switch (fileType) {
      case FileType.Graphic:
        return "file__delete_button_file_type_image"
      case FileType.PDF:
        return "file__delete_button_file_type_pdf"
      case FileType.Video:
        return "file__delete_button_file_type_video"
      default:
        return "file__delete_button_file_type_video"
    }
  }
}
export const getTitleKey = (titleKey: LucaI18nLangKey | undefined, fileType: FileType) => {
  if (titleKey) {
    return titleKey
  } else {
    switch (fileType) {
      case FileType.Graphic:
        return "file__edit_title_file_type_image"
      case FileType.PDF:
        return "file__edit_title_file_type_pdf"
      case FileType.Video:
        return "file__edit_title_file_type_video"
      default:
        return "file__edit_title_file_type_video"
    }
  }
}
