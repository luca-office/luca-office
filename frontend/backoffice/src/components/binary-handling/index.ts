export * from "./binary-form-placeholder/binary-form-placeholder"
export * from "./binary-update-modal/binary-entry"
export * from "./binary-update-modal/binary-update-modal"
export * from "./detail-view-binary/detail-view-binary"
export * from "./image-upload-modal/image-upload-modal"
