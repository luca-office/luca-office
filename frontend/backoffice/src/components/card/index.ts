// special backoffice card elements
export * from "./creation-card"
export * from "./duration-info/card-duration-info"
export * from "./tabbed-card/__mocks__/card-tabs.mock"
export * from "./tabbed-card/tabbed-card"
