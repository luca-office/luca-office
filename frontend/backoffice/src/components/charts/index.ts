export * from "./donut-chart/donut-chart-with-legend"
