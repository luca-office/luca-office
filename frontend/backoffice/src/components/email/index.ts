export * from "./email-body-meta-entry/email-body-meta-entry"
export * from "./email-body-setting-selection/email-body-setting-selection"
export * from "./email-content/email-content"
export * from "./email-directories-list/email-directories-list"
export * from "./email-header/email-header"
export * from "./email-unread-indicator/email-unread-indicator"
