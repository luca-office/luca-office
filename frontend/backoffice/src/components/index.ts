export * from "./binary-handling"
export * from "./card"
export * from "./card-overview/card-overview"
export * from "./charts"
export * from "./creation-bar/creation-bar"
export * from "./creation-status/creation-status"
export * from "./editable-multiline-textarea/editable-multiline-textarea"
export * from "./editing-status-indicator/editing-status-indicator"
export * from "./email"
export * from "./information-entry/information-entry"
export * from "./inline-editable-header/inline-editable-header"
export * from "./inline-editable-textarea/inline-editable-textarea"
export * from "./markdown-text/markdown-text"
export * from "./markdown-text-field/markdown-text-field"
export * from "./markdown-text-update-modal/markdown-text-update-modal"
export * from "./meta-entry/meta-entry"
export * from "./meta-entry-tags/meta-entry-tags"
export * from "./overlay-edit-field/overlay-edit-field"
export * from "./rating-progress-card/rating-progress-card"
export * from "./rating-progress-card/utils"
export * from "./resort-modal/resort-modal"
export * from "./sub-header-detail/sub-header-detail"
export * from "./sub-header-detail/sub-header-detail-container"
export * from "./sub-header-filter/sub-header-filter"
export * from "./table-action-bar/table-action-bar"
export * from "./upload-file-modal/upload-file-modal"
