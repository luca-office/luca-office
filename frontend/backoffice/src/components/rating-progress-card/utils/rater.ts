import {Rating, UserAccount} from "shared/models"
import {exists, isDefined} from "shared/utils"

export interface RaterCounts {
  readonly ratersInProgressCount: number
  readonly finishedRatersCount: number
  readonly ratings: Rating[]
}

interface RatingsByRater {
  readonly [raterId: string]: Rating[]
}

export const getRaterCounts = (allRatings: Rating[], raters: UserAccount[]): RaterCounts => {
  const ratings = allRatings.filter(rating => !rating.isFinalScore)
  const ratingsByRaters = ratings.reduce(
    (accumulator, rating) => ({
      ...accumulator,
      [rating.userAccountId]: [...(accumulator[rating.userAccountId] ?? []), rating]
    }),
    {} as RatingsByRater
  )

  const ratersCount = raters.length
  const finishedRatersCount = Object.keys(ratingsByRaters).filter(raterId =>
    exists(rating => isDefined(rating.finalizedAt), ratingsByRaters[raterId] ?? [])
  ).length
  const ratersInProgressCount = ratersCount - finishedRatersCount

  return {ratersInProgressCount, finishedRatersCount, ratings}
}
