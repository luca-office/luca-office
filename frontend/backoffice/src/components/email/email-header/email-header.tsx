import {css} from "@emotion/react"
import partial from "lodash-es/partial"
import * as React from "react"
import {DeleteOrArchiveEntityButton, Icon, Tooltip} from "shared/components"
import {EmailDirectory} from "shared/graphql/generated/globalTypes"
import {useDeleteEmail} from "shared/graphql/hooks"
import {Email} from "shared/models"
import {boxHeightMedium, Flex, inputHeight, spacingSmall} from "shared/styles"
import {useLucaTranslation} from "shared/translations"
import {getDirectoryIcon, getReceptionDelayLabel, Option} from "shared/utils"

export interface EmailHeaderProps {
  readonly actionsDisabled: boolean
  readonly email: Option<Email>
  readonly emailLoading: boolean
  readonly isIntroductionEmail: boolean
  readonly isInterventionEmail: boolean
  readonly hasIntervention: boolean
  readonly onDelete: () => void
}

export const EmailHeader: React.FC<EmailHeaderProps> = ({
  actionsDisabled,
  email,
  emailLoading,
  isIntroductionEmail,
  isInterventionEmail,
  hasIntervention,
  onDelete
}) => {
  const {t} = useLucaTranslation()

  const senderOrRecipient = email
    .map(
      mail =>
        `${mail.sender ?? mail.recipient} (${
          isIntroductionEmail
            ? t("email__introduction_email_label")
            : isInterventionEmail
            ? t("email__intervention_email_label")
            : getReceptionDelayLabel(t, mail.receptionDelayInSeconds)
        })`
    )
    .getOrElse(t("email__placeholder_not_found_title"))

  return (
    <div css={styles.wrapper}>
      <Icon
        customStyles={styles.directoryIcon}
        name={getDirectoryIcon(email.map(mail => mail.directory).getOrElse(EmailDirectory.Inbox))}
      />

      <div css={styles.senderOrRecipient}>{emailLoading ? t("email__placeholder_loading") : senderOrRecipient}</div>

      {!emailLoading &&
        email
          .map(mail => (
            <Tooltip
              title={t(
                hasIntervention
                  ? "email__intervention_email_delete_tooltip_email_has_intervention_error"
                  : "email__intervention_email_delete_tooltip"
              )}
              inactive={!isInterventionEmail && !hasIntervention}>
              <DeleteOrArchiveEntityButton
                useDeleteHook={partial(useDeleteEmail, mail.scenarioId)}
                entityId={mail.id}
                titleKey={isIntroductionEmail ? "email__title_delete_introduction" : undefined}
                onSuccess={onDelete}
                disabled={isIntroductionEmail || isInterventionEmail || hasIntervention || actionsDisabled}
              />
            </Tooltip>
          ))
          .orNull()}
    </div>
  )
}

const Size = {
  directoryIcon: 16
}

const styles = {
  wrapper: css(Flex.row, {
    height: boxHeightMedium
  }),
  directoryIcon: css({
    width: Size.directoryIcon,
    height: Size.directoryIcon,
    minWidth: Size.directoryIcon,
    marginLeft: spacingSmall,
    marginRight: spacingSmall
  }),
  senderOrRecipient: css({
    width: `calc(100% - ${2 * spacingSmall + Size.directoryIcon + inputHeight}px)`,
    marginRight: spacingSmall
  })
}
