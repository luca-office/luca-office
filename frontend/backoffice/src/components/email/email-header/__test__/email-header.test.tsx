import {render, screen} from "@testing-library/react"
import {shallow} from "enzyme"
import * as React from "react"
import {act} from "react-dom/test-utils"
import {act as actRenderer, create} from "react-test-renderer"
import {DeleteOrArchiveEntityButton, Icon} from "shared/components"
import * as useDeleteEmailHook from "shared/graphql/hooks/mutations/email/delete-email"
import {DeleteEntityHook} from "shared/models"
import {Option} from "shared/utils"
import wait from "waait"
import {emailsMock} from "../../../../modules/scenario-emails/emails/hooks/__mocks__/emails.mock"
import {EmailHeader, EmailHeaderProps} from "../email-header"

const defaultProps: EmailHeaderProps = {
  isIntroductionEmail: false,
  isInterventionEmail: false,
  email: Option.of(emailsMock[0]),
  emailLoading: false,
  onDelete: jest.fn(),
  actionsDisabled: false,
  hasIntervention: false
}

const deleteEntityHookValuesDefault: DeleteEntityHook = {
  deleteEntity: jest.fn(() => Promise.resolve()),
  deleteEntityLoading: false
}

const deleteEmailStateSpy = jest.spyOn(useDeleteEmailHook, "useDeleteEmail")

describe("email-header", () => {
  it("renders correctly", async () => {
    deleteEmailStateSpy.mockReturnValue(deleteEntityHookValuesDefault)
    const component = create(<EmailHeader {...defaultProps} />)
    await actRenderer(() => wait(0))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("renders correctly (intro)", async () => {
    deleteEmailStateSpy.mockReturnValue(deleteEntityHookValuesDefault)
    const component = create(<EmailHeader {...defaultProps} isIntroductionEmail={true} />)
    await act(() => wait(0))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("has correct structure", async () => {
    const tree = shallow(<EmailHeader {...defaultProps} />)

    await act(() => wait(0))

    expect(tree.find(Icon)).toHaveLength(1)
    expect(tree.find(DeleteOrArchiveEntityButton)).toHaveLength(1)
  })
  it("has correct structure (loading)", async () => {
    const tree = shallow(<EmailHeader {...{...defaultProps, emailLoading: true, email: Option.none()}} />)
    render(<EmailHeader {...{...defaultProps, emailLoading: true, email: Option.none()}} />)

    await act(() => wait(0))

    expect(tree.find(Icon)).toHaveLength(1)
    expect(tree.find(DeleteOrArchiveEntityButton)).toHaveLength(0)
    expect(screen.queryAllByText("email__placeholder_loading")).toHaveLength(1)
  })
  it("disables delete for introduction mail", async () => {
    const tree = shallow(<EmailHeader {...{...defaultProps, isIntroductionEmail: true}} />)

    await act(() => wait(0))

    expect(tree.find(DeleteOrArchiveEntityButton).prop("disabled")).toBeTruthy()
  })
  it("disables delete for readonly", async () => {
    const tree = shallow(<EmailHeader {...{...defaultProps, actionsDisabled: true}} />)

    await act(() => wait(0))

    expect(tree.find(DeleteOrArchiveEntityButton).prop("disabled")).toBeTruthy()
  })
})
