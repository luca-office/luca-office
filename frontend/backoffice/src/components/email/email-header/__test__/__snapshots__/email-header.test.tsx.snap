// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email-header renders correctly (intro) 1`] = `
<div
  className="css-5bayba-EmailHeader"
>
  <div
    className="css-1vyjarx-Icon"
    dangerouslySetInnerHTML={
      {
        "__html": {},
      }
    }
  />
  <div
    className="css-1ff2vdd-EmailHeader"
  >
    sender@cap3.de (email__introduction_email_label)
  </div>
  <div
    className="css-0"
  >
    <button
      className="css-hnekdl-Button"
      disabled={true}
      onClick={[Function]}
      title="email__title_delete_introduction"
    >
      <div
        className="css-95j2hm-Button"
      >
        <div
          className="css-mj8vo1-Icon"
          dangerouslySetInnerHTML={
            {
              "__html": {},
            }
          }
        />
      </div>
    </button>
  </div>
</div>
`;

exports[`email-header renders correctly 1`] = `
<div
  className="css-5bayba-EmailHeader"
>
  <div
    className="css-1vyjarx-Icon"
    dangerouslySetInnerHTML={
      {
        "__html": {},
      }
    }
  />
  <div
    className="css-1ff2vdd-EmailHeader"
  >
    sender@cap3.de (unit__just_now)
  </div>
  <div
    className="css-0"
  >
    <button
      className="css-hnekdl-Button"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="css-95j2hm-Button"
      >
        <div
          className="css-mj8vo1-Icon"
          dangerouslySetInnerHTML={
            {
              "__html": {},
            }
          }
        />
      </div>
    </button>
  </div>
</div>
`;
